msteams-platform/tabs/how-to/dashboard-teams-toolkit.md

```diff
@@ -0,0 +1,991 @@
+              <Spinner label="Loading..." labelPosition="below" />
+            </div>
+          ) : (
+            <div style={bodyContentStyle()}>
+              ...
+            </div>
+          )}
+        </>
+      );
+    }
+    ```
+
+1. **Hide the footer button if the data is loading**
+
+    Following is an example of footer button:
+
+    ```typescript
+    footerContent(): JSX.Element | undefined {
+      if (this.state.loading === false) {
+        return (
+          <Button
+            ...
+          </Button>
+        );
+      }
+    }
+    ```
+
+1. **Update the state reference**
+
+    Update the state reference in the widget file to use the new state type and update the state in the `getData` method to set the `loading` property to `false` after the data is loaded.
+
+    Now, the loading spinner is shown while the data is loading. When the data is loaded, the loading spinner is hidden, and the list data and footer button are shown.
+
+    :::image type="content" source="../../assets/images/sbs-create-a-new-dashboard/spinner.gif" alt-text="Graphical representation shows the loading spinner while the data is loading.":::
+
+### How to handle empty state
+
+You can display a specific content in your widget when the data is empty. To do so, you need to modify the `bodyContent` method in your widget file to adopt different states of the data. The following example shows how to display an empty image when the data of `ListWidget` is empty.
+
+```typescript
+bodyContent(): JSX.Element | undefined {
+    let hasData = this.state.data && this.state.data.length > 0;
+    return (
+      <div style={bodyContentStyle()}>
+        {hasData ? (
+          <>
+            {this.state.data?.map((t: ListModel) => {
+              ...
+            })}
+          </>
+        ) : (
+          <div
+            style={{
+              display: "grid",
+              gap: "1rem",
+              justifyContent: "center",
+              alignContent: "center",
+            }}
+          >
+            <Image src="empty-default.svg" height="150px" />
+            <Text align="center">No data</Text>
+          </div>
+        )}
+      </div>
+    );
+  }
+```
+
+You can use a similar approach to remove the footer content of your widget when the data is empty.
+
+```typescript
+footerContent(): JSX.Element | undefined {
+    let hasData = this.state.data && this.state.data.length > 0;
+    if (hasData) {
+      return (
+        <Button
+          ...
+        </Button>
+      );
+    }
+  }
+```
+
+  Following is an example of list widget when the data is empty:
+
+  :::image type="content" source="../../assets/images/sbs-create-a-new-dashboard/no-data.png" alt-text="Screenshot showing no data in the list.":::
+
+### How to refresh data as scheduled
+
+The following example shows how to display real-time data in a widget. The widget displays the current time and updates.
+
+```typescript
```

Comment:
It may be better if the language of this code snippet is adjusted to `tsx`.

Reply:
Updated to `tsx`